odoo/models.py

```diff
@@ -6811,6 +6814,75 @@ def _invalidate_cache(self, fnames=None, ids=None):
                 spec.append((invf, None))
         self.env.cache.invalidate(spec)
 
+    @contextlib.contextmanager
+    def modifying_recordset(self, *, outputs: Iterable[str], inputs: Iterable[str] = (), discard_inputs: bool = False) -> Iterator[SimpleNamespace]:
```

Comment:
I think this just adds complexity, you already have `invalidate*` and `flush*` functions and when building SQL queries, you can flush fields that you are querying.

So I don't this this API is needed. Maybe we could accept an iterable for the `to_flush` argument of `SQL`.

@rco-odoo What's your opinion?

Reply:
Would it not make sense to have this as part of the `execute_query` API? (by allowing to provide additional inputs/outputs by model name if we don't use `_fields_to_sql`)

I believe it is easier to use than a context manager for many people, and making the difference between `_recordset` and `_model` seems totally overkill since we didn't even care about invalidating before. (I know it is technically not correct, but there is no need to overengineer something for a problem that doesn't exist in practice.

Reply:
I would simply not deal with `modified`.
If the updates are done in SQL there are 3 main reasons:
* too many lines are updated, we can't bother with the ORM triggers etc
* we want to avoid ORM triggers on purpose
* the query needs some "advanced" SQL (i.e. `FOR UPDATE`), it means we sometimes don't even care if it is updated at all; it's just some best guess/try value.

At least for all queries patched in `account`, we don't want any of that.
For everything in mail, AFAIK we also only need very fast updates and don't care about all the triggers etc.

For the few other queries, everything was already managed correctly.

Reply:
Also, you could say that if there are outputs, then the query needs to be `RETURNING` the modifieds ids, and call modified on it

Reply:
> > If people don't know they should flush/invalidate before running SQL code and that they should call modified after they did an UPDATE in SQL, they probably won't know they should use this new method.
> 
> It's much easier to learn that you need to call a single wrapper with clear names than three different methods with subtle semantics.
> 
If someone does not understand the semantics of these methods, they probably should not be executing SQL directly. 

> > now you have 2 ways of doing things.
> 
> That is the purpose of higher level helpers. That "criticism" applies significantly more heavily to `execute_query`.
> 
Indeed, but maybe one day we could mask env.cr.

> > In the current state of things, you can use env.execute_query(SQL(...)) that will take care of flushing the inputs.
> 
> Which is far from sufficient for updates, hence the original need for this PR.
> 
Yes you still need to call modified 

> > Fictive example:
> 
> I like that your fictive example is completely dishonest, that's a great point of comparison:
> 
It's not, the example is equivalent to either call, not both. 

> ```python
> with self.modifiying_recordset(inputs=inputs, outputs=outputs):
>     self.env.execute_query(...)
> ```
> 
> expands to this:
> 
> ```python
> self.flush_recordset(inputs)
> self.env.execute_query(...)
> self.invalidate_recordset(outputs)
> self.modified(outputs)
> ```
> 
> and it _requires_ properly labelling inputs and outputs, which the latter does not because those are not the concepts it uses. As can, again, be seen in this PR.
> 
> > in case someone would access these values inside the block before the query is executed.
> 
Someone can do not that.
You can write with, then use an attribute to inject in the query. It's not done in this PR, but is a possible usage.